### PR TITLE
Bug 1816904: Ensure the Installer checks for the correct number of compute replicas before exiting

### PR DESCRIPTION
**Pull request by kirankt**
This PR ensures that the installer prints out an non-fatal error message before exiting if the number of worker replicas doesn't match the replicas requested in the install config.

These errors happen mostly in the baremetal platform where nodes might take a longer to finish introspection or enter an error state before which the installer might have exited leaving with incorrect compute/worker replicas. Since we ask the user for number of replicas in the install config, it makes sense to at least see to it that the replicas have been successfully deployed before exiting.

### CODE REVIEW

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -489,6 +521,11 @@ func logComplete(directory, consoleURL string) error {
 	logrus.Infof("To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=%s'", kubeconfig)
 	logrus.Infof("Access the OpenShift web-console here: %s", consoleURL)
 	logrus.Infof("Login to the console with user: %q, and password: %q", "kubeadmin", pw)
+
+	if err = checkComputeReplicas(config); err != nil {
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go:525]: don't put this as part of the logComplete. i don't think it belongs here.. we should probably just do that outside this function (before)

File: cmd/openshift-install/create.go [hunk 2]
```diff
@@ -473,8 +473,40 @@ func waitForConsole(ctx context.Context, config *rest.Config) (string, error) {
...
+	}
+
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get node list")
+	}
+
+	var computeReplicas, controlPlaneReplicas int64
+	if assetStore, err := assetstore.NewStore(rootOpts.dir); err == nil {
+		if installConfig, err := assetStore.Load(&installconfig.InstallConfig{}); err == nil && installConfig != nil {
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go:490]: probably use Fetch like https://github.com/openshift/installer/blob/fcf8f4517c79a65ca4f14f1aa5af48f66cea3df9/cmd/openshift-install/gather.go#L97-L101

File: cmd/openshift-install/create.go [hunk 3]
```diff
@@ -473,8 +473,40 @@ func waitForConsole(ctx context.Context, config *rest.Config) (string, error) {
...
+			if installConfig.(*installconfig.InstallConfig).Config.Platform.Name() == baremetal.Name {
```
Comment (abhinavdahiya) [inline cmd/openshift-install/create.go:491]:
```go
if ic, ok := installConfig.(*installconfig.InstallConfig); ok && <> {
  ...
  for _, compute := ic.Config.Compute {
  }
  ...
}
```

Comment (kirankt) [reply cmd/openshift-install/create.go]: Done. Thanks.

File: cmd/openshift-install/create.go (hunk 1 above)
Comment (kirankt) [reply cmd/openshift-install/create.go]: +1. Moved it into waitForInstallComplete and converted the error into a warning.

**Comment by kirankt** [comment]
> Personally i don't think it is the installer's responsibilty to check if the compute nodes are running or how many,
> 
> We should probably work with machine-api-operator to track a condition on it's cluster operator that reports the machines are that haven't yet become nodes ... the installer can then provide that condition as part of the error message when the cluster complete failed.
> 

Hi. Thanks for the feedback. I was wondering if there is an example of how we currently use the cluster operator in the installer. 

> If the cluster complete is done, worker nodes that are still provisioning or haven't provisioned means that they were not required for install-complete and hence it's weird to prompt that in installer, the cluster already has an alert for such machines in those cases.

Instead of an error, can we just print a warning before we call logcomplete? Instead of an error, I'm currently just printing (in the newest commit) a warning, so its not too disruptive. WDYT?

**Comment by stbenjam** [comment]
> Personally i don't think it is the installer's responsibilty to check if the compute nodes are running or how many,

Everyone points to everyone else as responsibility for handling this :man_shrugging: 

We've previously discussed this on https://github.com/openshift/enhancements/pull/328#discussion_r427560374

**Comment by kirankt** [comment]
@abhinavdahiya What do you think of the new changes. We're just printing a warning message before exiting. IMO its pretty harmless and will not really affect other platforms.

**Comment by kirankt** [comment]
@abhinavdahiya . Ideally we would like to exit with a non-zero return code, rather than a warning, which is what this PR is currently doing. Also, I've ensured that this affects only the baremetal platform. PTAL and let me know what you think.
/cc @hardys

**Comment by abhinavdahiya** [comment]
> @abhinavdahiya . Ideally we would like to exit with a non-zero return code, rather than a warning, which is what this PR is currently doing. Also, I've ensured that this affects only the baremetal platform. PTAL and let me know what you think.
> /cc @hardys

even for baremetal my concerns are still the same from https://github.com/openshift/installer/pull/4071#pullrequestreview-470717673 and none of it are fundamentally answered.

How do you ensure our figure out if the compute node that is missing or still not running actually required for install-complete signal? Marking installs as failed in these cases imo is inaccurate and should be avoided.

If you can update this PR to that constraint, more than happy to have the installer run this check for install complete.

secondly the people also use this code-path for UPI clusters and expecting/requiring those customers to match the install config replica count seems like an unnecessary requirement.

**Comment by kirankt** [comment]
> > @abhinavdahiya . Ideally we would like to exit with a non-zero return code, rather than a warning, which is what this PR is currently doing. Also, I've ensured that this affects only the baremetal platform. PTAL and let me know what you think.
> > /cc @hardys
> 
> even for baremetal my concerns are still the same from [#4071 (review)](https://github.com/openshift/installer/pull/4071#pullrequestreview-470717673) and none of it are fundamentally answered.
> 
> How do you ensure our figure out if the compute node that is missing or still not running actually required for install-complete signal? Marking installs as failed in these cases imo is inaccurate and should be avoided.

Thanks, @abhinavdahiya  for taking time to re-review.

I understand your concern. It makes no sense to error out an otherwise perfectly installed cluster. What about how this PR stands now? Currently we're only printing a warning message about compute replicas. Can we at least do this (for BM platform only) and then continue to the normal exit? I think I'll be OK with this compromise. These bugs are being reported a lot and doesn't seem to be a corner case. At least if the warning message exists, we can inform the end user about it and they can take action on day-2.

> 
> If you can update this PR to that constraint, more than happy to have the installer run this check for install complete.
> 
> secondly the people also use this code-path for UPI clusters and expecting/requiring those customers to match the install config replica count seems like an unnecessary requirement.

FWIW, UPI uses platform=none for their installs and AFAIK it should not affect them. :-)

**Comment by stbenjam** [comment]
I moved this bug out of the baremetal installer queue and back to the core installer team, this bug was reported in March and we've been playing a game of musical chairs with it. I don't think we're any closer to a resolution on it, and IMHO it really should be a fix for every platform in OCP.

Let's close this PR. Feel free to reopen if there's a path forward that fixes it for everyone.

/close

**Pull request closed without merging**